wpilibj/src/main/java/edu/wpi/first/wpilibj/Ultrasonic.java

```diff
@@ -68,19 +70,29 @@
   private static class UltrasonicChecker extends Thread {
     @Override
     public synchronized void run() {
-      Ultrasonic ultrasonic = null;
+      int sensorIndex = 0;
+      Ultrasonic ultrasonic;
       while (m_automaticEnabled) {
-        if (ultrasonic == null) {
-          ultrasonic = m_firstSensor;
-        }
-        if (ultrasonic == null) {
-          return;
+        //lock list to ensure deletion doesn't occur between empty check and retrieving sensor
+        synchronized (m_sensors) {
+          if (m_sensors.isEmpty()) {
+            return;
+          }
+          if (sensorIndex >= m_sensors.size()) {
+            sensorIndex = m_sensors.size() - 1;
+          }
+          ultrasonic = m_sensors.get(sensorIndex);
         }
         if (ultrasonic.isEnabled()) {
           // Do the ping
           ultrasonic.m_pingChannel.pulse(kPingTime);
         }
```

Comment:
There is still a race condition here.  Once we exit the synchronized block, ultrasonic could be closed and cause a null pointer exception on m_pingChannel.  What you should do is move the entire if block to a private synchronized void doPing() method and then call that method here.  doPing will also need to check if m_pingChannel is not null.